### PR TITLE
Change Controller::__construct() to accept an array as its argument

### PR DESCRIPTION
**Pull request by rchavik**
https://github.com/cakephp/cakephp/pull/7301#discussion_r37954417

### CODE REVIEW

File: src/Controller/Controller.php [hunk 1]
```diff
@@ -222,15 +222,39 @@ class Controller implements EventListenerInterface, EventDispatcherInterface
...
-     * @param \Cake\Event\EventManager|null $eventManager The event manager. Defaults to a new instance.
-     * @param \Cake\Controller\ComponentRegistry|null $components The component registry. Defaults to a new instance.
+     * - response: instance of \Cake\Network\Response object for this controller.
+     * - name: Override the name useful in testing when using mocks.
+     * - eventManager: Instance of \Cake\Event\EventManager. Defaults to a new instance.
+     * - components: Instance of \Cake\Controller\ComponentRegistry. Defaults to a new instance.
+     * @param array $config Options
      */
-    public function __construct(Request $request = null, Response $response = null, $name = null, $eventManager = null, $components = null)
+    public function __construct($options = [])
```
Comment (markstory) [inline src/Controller/Controller.php]: This adds a bunch of code and makes type-hinting harder, what is the benefit?

Comment (rchavik) [inline src/Controller/Controller.php]: I can't exactly where I say this: Future Proof APIs, Options parameters make growing easy. 

:)

**Comment by robertpustulka** [comment]
I'm :-1: on this. It makes harder to work with reflection.

**Comment by ADmad** [comment]
Given that the developer almost never needs to instantiate controller "by hand" I too don't see the benefit of accepting args as array.

**Comment by lorenzo** [comment]
On the fence here, I think that if we had a compelling reason to do this in the future we could just do it. So far I don't see a big benefit, on the other had it will break my dependency injection plugin as no typehints will be found :(

**Comment by markstory** [comment]
The bag of stuff parameter works ok when we have a servicelocator/factory like TableRegistry. But we don't have one of those for controllers.

**Comment by rchavik** [comment]
It feels ugly. Especially the `name` argument, that looks really out of place there.  A special arg for testing? Feels like 'cheating'.

Pity php has no named params. :(

**Comment by lorenzo** [comment]
Named params would make everything so much better :D

**Comment by dereuromark** [comment]
Indeed.
Well, seems like we should hold off on that change and alike until 4.x.

**Comment by rchavik** [comment]
Closing :(

**Comment by robertpustulka** [comment]
A class should never have too many dependencies. 3-4 are OK, 5 are max. If you need more then your class is badly designed. I don't see why we should make such future-proof changes. If the need emerges, we'll find another, more elegant way to deal with many dependencies.

**Pull request closed without merging**